### ISSUE TITLE
final incompatible with haxe.Unserializer

### ISSUE THREAD

**Issue by ckentgeorge** [HaxeFoundation/haxe#8263]
There's a frustrating intersection between `final` variables only being settable at declaration or in the constructor, and `haxe.Unserializer` creating empty instances, requiring initialization outside of the class constructor. This essentially precludes the use of `final` on any class that we wish to use serialization with.

Maybe we could add some kind of metadata to indicate that initialization should be allowed from another function, not just `new()`. Maybe there's another way to tackle this.

I'm testing on Haxe 4.0.0-rc.2. Here's what I'm trying to do:
```haxe
class Main {
    static function main() {
        var unserializer = new Unserializer(data);
    }
}

class Tweenable {
    public function new() { }
}

class Thing {
    public final x:Tweenable;

    public function new() {
        // This would work:
        x = new Tweenable();
        initialize();
    }

    private function initialize():Void {
        // But we need this, and it doesn't
        x = new Tweenable();
    }

    @:allowInitialization
    @:keep
    public  function hxUnserialize(unserializer:Unserializer) {
        initialize();
        // ...
    }
}
```

**Comment by lublak** [comment]
Unserializer use the reflection api.
On which target did you test it? 

(Your direct access can not work.
You should test it with Reflect.setField like the Unserializer)

**Comment by ckentgeorge** [comment]
I think you're right, but it uses `Reflect.createEmptyInstance()` which skips the constructor and member initialization.
`Reflect.setField()` seems a bit hacky for something implied by the Haxe library itself. I also have performance concerns about it, but yes, I suppose it would work.
I'm targeting C++, Java, and Javascript

**Comment by RealyUniqueName** [comment]
I guess It's not about `haxe.Unserializer`, but about some custom serialization process.
Maybe we could add something like `@:hack` to ignore `final` semantics in certain situations.
@Simn ?

**Comment by ckentgeorge** [comment]
@RealyUniqueName, I think you're right. The need for some kind of exception to the rule is more directly implied by the existence of `Reflect.createEmptyInstance()`, since using it would preclude the use of `final` without using reflection.

**Comment by back2dos** [comment]
Hmm, somehow the desired behavior seems wrong to me. There's a field that's declared final, but a public method that can cause it's modification.

The cleanest solution we be to make `hxUnserialize` static (so you can construct an instance from there or even using pooling and what not), which makes a whole lot more sense in general, but is a breaking change.

**Comment by ckentgeorge** [comment]
@back2dos, I surmise the reason for going the empty instance route, was to avoid the inefficiency of double-initialization. While your approach would fix my problem, it wouldn't fix the general case -- unserialization needing it's own initialization path separate from the constructor. `@:initializer`?

**Comment by lublak** [comment]
@ckentgeorge i think it could be usefull. But i think its very complex to create it.
As I heard, final should also be used native.
This means we can not have a method to inialize a object.
The only solution I would just know would be an extra static initializer.
Or multiple constructors

```haxe
@:emptyInstance
public static function createOwnEmptyInstance(tthis:Test, data:String) {
    tthis.anyFinalValue = data;
    return tthis;
}
```

For information:
java/c#/c++:
an extra construtor for any @:init function or multiple haxe constructors

**Comment by ncannasse** [comment]
Maybe we could use the new @:byPassAccessor to allow access to final vars (ping @nadako)

**Comment by Gama11** [comment]
That was already discussed briefly: https://github.com/HaxeFoundation/haxe/pull/8259#issuecomment-490416057

**Comment by lublak** [comment]
I think alternative we can have a :bypassFinal to force a normal variable?

**Comment by nadako** [comment]
Yeah this is a valid use-case. I wonder if generating target-native `final` makes any real difference besides being more nice...

**Comment by ncannasse** [comment]
I don't think we should generate target native final as our final rules might not match target one anyway.

**Comment by back2dos** [comment]
> @back2dos, I surmise the reason for going the empty instance route, was to avoid the inefficiency of double-initialization. While your approach would fix my problem, it wouldn't fix the general case -- unserialization needing it's own initialization path separate from the constructor. @:initializer?

You can (and I'd say you should) factor the code paths properly. This is (if I get your point) roughly the problem you describe:

```haxe
class Foo {
  final a:Int;
  public function new(a) {
    this.a = a;
    doSomethingYouWantToSkipWhenUnserializing();
  }

  function doSomethingYouWantToSkipWhenUnserializing() { /* whatever */ }

  @:keep static public function hxUnserialize(unserializer:Unserializer) 
    return new Foo(unserializer.unserialize());// oh no, we ran that thing we didn't wanna run
}
```

Might as well be this:

```haxe
class Foo {
  final a:Int;
  function new(a) {
    this.a = a;
  }
  static public function create(a) {
    var foo = new Foo(a);
    foo.doSomethingYouWantToSkipWhenUnserializing();
    return foo;
  }
  function doSomethingYouWantToSkipWhenUnserializing() { /* whatever */ }

  @:keep static public function hxUnserialize(unserializer:Unserializer) 
    return new Foo(unserializer.unserialize());
}
```

That said, there's an important case for `Type.createEmptyInstance()` approach that I've overlooked: It allows the unserializer to put the instance into cache before unserialization proceeds, thereby allowing for circular references to be handled properly.

----

I think a pretty decent solution would be to have the unserializer also check for static `hxUnserialize` methods. You can then use the above pattern to deal with final fields. It may feel a bit like a roundabout way, but it does fully honor the semantics of `final`.

**Comment by ckentgeorge** [comment]
I think I like that deserialization has its own initialization path so it's not necessary to use serializable classes that run into the final issue differently, i.e. `create()` instead of directly using the constructor.

I'm actually super curious if there is another reason, beyond serialization, why the `Reflect.createEmptyInstance()` function exits. Is it used for anything else? So far, serialization is the only thing I've run into that seems to benefit from this feature. If we don't like making exceptions for 'final', then maybe we should nix creating empty instances altogether.

**Comment by Simn** [comment]
I agree with @back2dos that `hxUnserialize` should be allowed to be static. It's a clean solution and doesn't require messing with `final` at all. I'm against adding a bypass for this because then we cannot reason about `final` at all and it becomes useless as far as optimization is concerned.

**Comment by ckentgeorge** [comment]
Can we kill the whole notion of `createEmptyInstance()` while we're at it?

**Comment by back2dos** [comment]
Nope, that's still required if you wish to unserialize cyclic object graphs ;)

**Comment by ckentgeorge** [comment]
Then this totally fails with my use-case. Either it's one or the other; both don't make sense. This solution would mean I can't use a cyclic object graph + final. This seems arbitrary and may lead developers down the path of unpleasant surprises.

**Comment by Gama11** [comment]
Well, `createEmptyInstance()` is reflection... You could also modify `final` fields with reflection.

**Comment by ckentgeorge** [comment]
Right, but part of the argument for not making exceptions, is wanting to propagate `final` all the way down to the target language. Will reflection still work in this case?

**Comment by Gama11** [comment]
That would depend on each specific target's runtime I imagine, I doubt Haxe could / should _guarantee_ that it works.

**Comment by ckentgeorge** [comment]
That is my suspicion as well. If this is true, then the final behavior becomes platform specific. This seems to be contrary to the goals of Haxe.

**Comment by Simn** [comment]
Reflection has always been target-specific to an extent.

Let's focus on the final deserialization issue please.

**Comment by back2dos** [comment]
> This solution would mean I can't use a cyclic object graph + final.

That's pretty much a given anyway. With anonymous objects, you can't have a cyclic immutable graph. The one way to get this done with classes is this:

```haxe
class A {
  final b:B;
  public function new() this.b = new B(this);
}

class B {
  final a:A;
  public function new(a) this.a = a;
}
```

This is quite the edge case (and it probably shouldn't even compile, because the reference to `this` in `A.new` escapes before the object is fully initialized). In general, you need mutability to make object graphs cyclic.

> If this is true, then the final behavior becomes platform specific. This seems to be contrary to the goals of Haxe.

The goal of `final` is to have a static guarantee that a value won't change and unless you totally bypass the type system, that guarantee is upheld across all platforms. This is consistent for practically all features of the type system. The cost of enforcing all compile time semantics at runtime would be enormous and the benefit would be quite minimal (at least under the assumption that the user knows what they're doing when using reflection/`untyped`/`Dynamic`).

**Comment by lublak** [comment]
I think final can not be native. So reflection should always work.
Why final is impossible:

```haxe
abstract TestString (String) {
    public inline function new(str:String) this = str;
    public inline function assing(str:String) this = str; 
}
clas Main {
    public final test:TestString;
    public function new() {
        test = new TestString("string");
    }
    public static function main() {
        var main = new Main();
        main.test.assing("otherstring"); // internal it creates something like this: main.test = "otherstring";
    }
}
```

I think haxe support "@:hack Allows extending classes marked as @:final".
Can also be used for hacking a field.

**Comment by ncannasse** [comment]
Let's get back to pragmatic mode:
- we should not generate target native final as our final rules will most likely not match target one
- Reflection would then work without any problem, as it actually should (which is another reason for no native final)
- so users can call Reflect.setField on final fields, but that's very unsafe way to do
- so let's allow them to do @:byPassAccessor on final fields - at their own risks - which means the compiler can still consider the field final when it comes to its own optimizations.

**Comment by Simn** [comment]
I don't really understand why you consider `@:byPassAccessor` to be any safer than `Reflect.setField` here. What exactly is the difference?

**Comment by RealyUniqueName** [comment]
It's not safier in terms of "finaliness", but safier in terms of types and better performance at runtime.

**Comment by Gama11** [comment]
Well, the former is still type-safe (apart from the `final`). `Reflect.setField` is completely `Dynamic` / stringly typed with its arguments.

**Comment by Simn** [comment]
I don't really see the point in introducing type-safe concepts with a "use at your own risk" warning label. That's a weird middleground that adds complexity for no benefit, because you cannot rely on it anyway...

**Comment by RealyUniqueName** [comment]
We already have `@:privateAccess`...

**Comment by Gama11** [comment]
Yeah, it's a bit strange.. At the very least it should maybe not be mixed with `@:bypassAccessor`, which is actually safe in its current form.

`@:privateAccess` is also safe to use, isn't it?

**Comment by back2dos** [comment]
So, which part of just allowing static hxUnserialize is not pragmatic @ncannasse ?

**Comment by ncannasse** [comment]
It's the design and spirit of Haxe of not trying to "protect developers against themselves", at least when they are explicit about what they want to do. So yes we need a way to set final fields.

We could allow static hxUnserialize _as well_ but that would only solve the very particular case of haxe.Serializer.

**Comment by back2dos** [comment]
> We could allow static hxUnserialize as well but that would only solve the very particular case of haxe.Serializer.

I guess you mean haxe.Unserializer? I have shown above how it would solve the use case in question. OP has agreed that the solution works for him.  You're looking to solve a non-issue via more escape hatches in the type system. Can we please actually be pragmatic and defer such measures until there's at least one person in need for it?

**Comment by ncannasse** [comment]
I don't think that fixing issues by changing a language library is the way to do.
If you really insist I can also open a similar issue about hxbit serialization to prove my point.

**Comment by ckentgeorge** [comment]
Just to clarify, I am using a cyclic object graph, and I would be greatly benefited by the use of final. If I can use reflection to set final properties reliably cross-target-language, then fine; I don't love it, but it works. If I had a vote, I'd vote in favor of not changing the reflection libraries, as I think that creating empty instances brings value in certain situations, and I think creating separate paths adds unnecessary complexity. I would also like to avoid the overhead of reflection, which is why I would vote in favor of using @:byPassAccessor over the reflection route.

**Comment by back2dos** [comment]
Weren't you proposing to axe `Type.createEmptyInstance()` just a while ago? Can you please show the immutable cyclic object graph though (because that has me quite intrigued)?

What puzzles me about this issue is the insistence on using `final` when at the same time not wanting its semantics, but since Nicolas seems to have his mind made up already, I withdraw. You kids have fun.

**Comment by ckentgeorge** [comment]
Hmm. I can see why you thought I was proposing the removal of `Type.createEmptyInstance()`, but that's not quite what I meant. I was trying to point out that if final doesn't work with the deserialization use-case of creating empty instances, then maybe creating empty instances shouldn't exist at all. It's one or the other, but both don't make sense, if we don't support initializing final members via this alternate path. I see deserialization as essentially leveraging an alternate construction/initialization path.

So, I'm not sure how I would succinctly demonstrate the cyclic nature of my object graph in its specifics, but I don't think there's anything particularly unique about it theoretically. The immutability is needed because we are exposing tweenable property handlers. I want to allow users to extend a class that has these tweenable properties, but make it hard for them to inadvertently replace the object directly. (Because of some operator overloading and abstract magic, we are actually running into this problem.)

**Comment by nadako** [comment]
So I'm working on the "shared game logic" thingy once again and I think I agree with @ncannasse's thoughts regarding `final` fields.

In my case I want deserialization to be completely hidden from the programmer and thus having to structure classes in a specific way like @back2dos suggests is undesirable. Of course, building these static constructors with a macro is an option, but I would like to preserve normal class instantiation syntax, so people can write their data classes and use it with a natural `new SomeClass()`.

Also I have a nice use-case for `final` in my data classes: I macro-generate read-only "view" classes with "reactive properties" and `final` can help figuring out whether I need to generate a reactive property or a normal property.

IMO deserialization in general is a very low-level business, and I think having some explicit hacks like `@:bypassAccessor` to support it is fine. I really appreciate how, unlike a lot of other languages, Haxe can implement deserialization in a way that does not require a programmer to write their classes in a specific way to make deserialization framework happy.

**Comment by Aurel300** [comment]
I also think supporting `static` unserialisation functions is the way to go. If you happen to be unserialising anything with a parent class, there is no way to call its constructor. E.g.:

```haxe
class Base {
  function new() trace("base class init");
}
class Child extends Base {
  public function new() super();
  function hxUnserialize(u) {
    // super(); // would throw an error
  }
}
```

So any initialisation that is done in the super-constructor either needs to be replicated, or moved around such that it is in a separate function, callable from the unserialisation function, which introduces further messiness if the parent class is using `final` as well.

**Comment by Aurel300** [comment]
It would be nice to solve this to make `asys` APIs nicer.

**Comment by RealyUniqueName** [comment]
After re-reading this discussion once again I think we should not introduce things to subvert `final`.
Final is final, it does its job. If you find yourself in need to bypass `final`, then you either need to refactor your code or switch to a var property. 
Var property also allows to keep things safe from users, but without restricting property holder in modification of it.
